engine/classes/Elgg/ActionsService.php

```diff
@@ -69,24 +76,32 @@ public function execute($action, $forwarder = "") {
 		}
 
 		/**
-		 * Complete the execution with a forward
-		 *
-		 * @param string $error_key Error message key
+		 * Prepare action response
 		 *
-		 * @throws \SecurityException
+		 * @param string $error_key   Error message key
+		 * @param int    $status_code HTTP status code
+		 * @return \Elgg\Http\ResponseBuilder
 		 */
-		$forward = function ($error_key = '') use ($action, $forwarder) {
+		$forward = function ($error_key = '', $status_code = null) use ($action, $forwarder) {
+			$forwarder = empty($forwarder) ? REFERER : $forwarder;
 			if ($error_key) {
 				$msg = _elgg_services()->translator->translate($error_key, [$action]);
 				_elgg_services()->systemMessages->addErrorMessage($msg);
+				if (!$status_code) {
+					$status_code = ELGG_HTTP_BAD_REQUEST;
+				}
 			}
 
+			if (!$status_code) {
+				$status_code = ELGG_HTTP_OK;
+			}
+			
 			$forwarder = empty($forwarder) ? REFERER : $forwarder;
-			forward($forwarder);
+			return _elgg_services()->responseFactory->redirect($forwarder, $status_code);
 		};
 
 		if (!isset($this->actions[$action])) {
-			$forward('actionundefined');
+			return $forward('actionundefined', ELGG_HTTP_NOT_IMPLEMENTED);
```

Comment:
Ajax calls with `elgg.ajax` receive 200 for all calls, except `ajax/view` and `ajax/form`.
For `elgg/Ajax`, I am hoping your latest PR will fix that.